Data delivery broken

This happened when ism-go was the source of the copy.
Python version did not cause this error.
Seems to be in the area of speedmap archive data.

```console
vscode ➜ /workspaces/racelogctl (main ✗) $ go run main.go event copy 50 --source-url ws://localhost:8091/ws -p abcdef
Using config file: /workspaces/racelogctl/.racelogctl.yml
Created event on target:
Id: 2 (Key: a176385b04601a74fc99d792d0da754e)
Name: Petite Lemans
Recorded: 2023-03-07 22:19 (racelogger: 0.4.4)
Track: Road Atlanta
Session begin: 0s 0
Session end: 0s 0
Race begin (UTC): 1970-01-01 00:00 (0)
2023/03/07 21:19:38 begin copy states
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
racelogctl/wamp.(*PublicClient).GetStates(0xc000206060, 0xc0002285c0?, 0x41d8d07f538ed409, 0xc0002429c0?)
        /workspaces/racelogctl/wamp/public_client.go:171 +0x465
racelogctl/cmd.copyStandardData({0xc000206060, 0xc000206158, 0x32, {0xc0002285c0, 0x20}})
        /workspaces/racelogctl/cmd/event_copy.go:176 +0x109
racelogctl/cmd.eventCopy(0x32)
        /workspaces/racelogctl/cmd/event_copy.go:151 +0x69a
racelogctl/cmd.glob..func4(0xcb9a40?, {0xc000122730?, 0x5?, 0x5?})
        /workspaces/racelogctl/cmd/event_copy.go:64 +0x7b
github.com/spf13/cobra.(*Command).execute(0xcb9a40, {0xc0001226e0, 0x5, 0x5})
        /go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:920 +0x847
github.com/spf13/cobra.(*Command).ExecuteC(0xcba000)
        /go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        /go/pkg/mod/github.com/spf13/cobra@v1.6.1/command.go:968
racelogctl/cmd.Execute()
        /workspaces/racelogctl/cmd/root.go:37 +0x25
main.main()
        /workspaces/racelogctl/main.go:10 +0x17
exit status 2
```

Frontend is not happy, too
![image](https://user-images.githubusercontent.com/41362745/226187757-35a1b273-fc58-40d5-8d9d-4e6dbaa42968.png)
![image](https://user-images.githubusercontent.com/41362745/226187796-afff79f8-376f-45b0-86ec-0edc5090258a.png)

--- a/.vscode/launch.json
+++ b/.vscode/launch.json
@@ -10,7 +10,7 @@
       "request": "launch",
       "mode": "auto",
       "program": "${workspaceFolder}/main.go",
-      "args": ["server", "--logFormat", "text", "--logLevel", "info"]
+      "args": ["server", "--logFormat", "text", "--logLevel", "debug"]
     },
     {
       "name": "Migrate",

--- a/pkg/repository/speedmap/speedmap_repos.go
+++ b/pkg/repository/speedmap/speedmap_repos.go
@@ -13,11 +13,11 @@ import (
 	"github.com/mpapenbr/iracelog-service-manager-go/pkg/repository"
 )
 
-func Create(conn repository.Querier, state *model.DbSpeedmap) error {
+func Create(conn repository.Querier, sm *model.DbSpeedmap) error {
 	_, err := conn.Exec(
 		context.Background(),
 		"insert into speedmap (event_id, data) values ($1,$2)",
-		state.EventID, state.Data)
+		sm.EventID, sm.Data)
 	if err != nil {
 		return err
 	}
@@ -76,16 +76,24 @@ func LoadRange(conn repository.Querier, eventID int, tsBegin float64, num int) (
 ) {
 	var rows pgx.Rows
 	if rows, err = conn.Query(context.Background(), `
-	select data from speedmap
+	select id,event_id,data from speedmap
     where event_id=$1 and (data->'timestamp')::numeric > $2
     order by (data->'timestamp')::numeric asc
     limit $3`,
 		eventID, tsBegin, num); err == nil {
-		ret, err = pgx.CollectRows[*model.SpeedmapData](rows,
-			func(row pgx.CollectableRow) (*model.SpeedmapData, error) {
-				return pgx.RowToAddrOfStructByPos[model.SpeedmapData](row)
+		tmp, collErr := pgx.CollectRows[*model.DbSpeedmap](rows,
+			func(row pgx.CollectableRow) (*model.DbSpeedmap, error) {
+				return pgx.RowToAddrOfStructByPos[model.DbSpeedmap](row)
 			})
-		return ret, err
+		if collErr == nil {
+			ret = make([]*model.SpeedmapData, len(tmp))
+			for i, v := range tmp {
+				ret[i] = &v.Data
+			}
+		} else {
+			ret = nil
+		}
+		return ret, collErr
 	} else {
 		return nil, err
 	}

--- a/pkg/repository/speedmap/speedmap_repos_test.go
+++ b/pkg/repository/speedmap/speedmap_repos_test.go
@@ -1,0 +1,77 @@
+//nolint:dupl,funlen,errcheck //ok for this test code
+package speedmap
+
+import (
+	"context"
+	"log"
+	"testing"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/mpapenbr/iracelog-service-manager-go/pkg/model"
+	"github.com/mpapenbr/iracelog-service-manager-go/pkg/repository/event"
+	tcpg "github.com/mpapenbr/iracelog-service-manager-go/testsupport/tcpostgres"
+)
+
+func initTestDb() *pgxpool.Pool {
+	pool := tcpg.SetupTestDb()
+	tcpg.ClearAllTables(pool)
+	return pool
+}
+
+func TestLoadRange(t *testing.T) {
+	db := initTestDb()
+	dbEvent := &model.DbEvent{
+		Name:        "Test",
+		Key:         "testKey",
+		Description: "myDescr",
+		Data:        model.EventData{},
+	}
+
+	err := pgx.BeginFunc(context.Background(), db, func(tx pgx.Tx) error {
+		e, err := event.Create(tx.Conn(), dbEvent)
+		for i := 1; i <= 3; i++ {
+			Create(tx.Conn(), &model.DbSpeedmap{
+				EventID: e.ID,
+				Data:    model.SpeedmapData{Type: i, Timestamp: float64(i * 10.0)},
+			})
+		}
+		return err
+	})
+	if err != nil {
+		log.Fatalf("createTestdata: %v\n", err)
+	}
+
+	type args struct {
+		tsBegin float64
+		num     int
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantRet []int
+		wantErr bool
+	}{
+		{name: "all", args: args{tsBegin: 0, num: 9}, wantRet: []int{1, 2, 3}},
+		{name: "none", args: args{tsBegin: 0, num: 0}, wantRet: []int{}},
+		{name: "oob-upper", args: args{tsBegin: 40.0, num: 10}, wantRet: []int{}},
+		{name: "last", args: args{tsBegin: 20.0, num: 10}, wantRet: []int{3}},
+		{name: "just two", args: args{tsBegin: 0.0, num: 2}, wantRet: []int{1, 2}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotRet, err := LoadRange(db, dbEvent.ID, tt.args.tsBegin, tt.args.num)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("LoadRange() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			extractTypes := make([]int, 0)
+			for _, v := range gotRet {
+				extractTypes = append(extractTypes, v.Type)
+			}
+			assert.Equal(t, tt.wantRet, extractTypes)
+		})
+	}
+}

--- a/pkg/repository/state/state_repos_test.go
+++ b/pkg/repository/state/state_repos_test.go
@@ -1,10 +1,26 @@
-//nolint:funlen,lll // ok for param tests
+//nolint:funlen,lll,errcheck // ok for param tests
 package state
 
 import (
+	"context"
+	"log"
 	"reflect"
 	"testing"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/mpapenbr/iracelog-service-manager-go/pkg/model"
+	"github.com/mpapenbr/iracelog-service-manager-go/pkg/repository/event"
+	tcpg "github.com/mpapenbr/iracelog-service-manager-go/testsupport/tcpostgres"
 )
+
+func initTestDb() *pgxpool.Pool {
+	pool := tcpg.SetupTestDb()
+	tcpg.ClearAllTables(pool)
+	return pool
+}
 
 func Test_computeCarChanges(t *testing.T) {
 	type args struct {
@@ -123,6 +139,75 @@ func Test_computeSessionChanges(t *testing.T) {
 			if got := computeSessionChanges(tt.args.ref, tt.args.cur); !reflect.DeepEqual(got, tt.want) {
 				t.Errorf("computeSessionChanges() = %v, want %v", got, tt.want)
 			}
+		})
+	}
+}
+
+func TestLoadByEventIdWithDelta(t *testing.T) {
+	db := initTestDb()
+	dbEvent := &model.DbEvent{
+		Name:        "Test",
+		Key:         "testKey",
+		Description: "myDescr",
+		Data:        model.EventData{},
+	}
+
+	err := pgx.BeginFunc(context.Background(), db, func(tx pgx.Tx) error {
+		e, err := event.Create(tx.Conn(), dbEvent)
+		for i := 1; i <= 3; i++ {
+			Create(tx.Conn(), &model.DbState{
+				EventID: e.ID,
+				Data: model.StateData{
+					Type: 1,
+					Payload: model.StatePayload{
+						Cars: [][]interface{}{
+							{i, 2},
+						},
+						Session: []any{i, 2},
+					}, Timestamp: float64(i * 10.0),
+				},
+			})
+		}
+		return err
+	})
+	if err != nil {
+		log.Fatalf("createTestdata: %v\n", err)
+	}
+
+	var type1 int = 1
+	type args struct {
+		startTS float64
+		num     int
+	}
+	tests := []struct {
+		name      string
+		args      args
+		wantRef   any
+		wantDelta []int
+		wantErr   bool
+	}{
+		{name: "all", args: args{startTS: 0, num: 10}, wantRef: &type1, wantDelta: []int{2, 2}},
+		{name: "none", args: args{startTS: 0, num: 0}, wantRef: nil, wantDelta: []int{}},
+		{name: "oob", args: args{startTS: 30, num: 10}, wantRef: nil, wantDelta: []int{}},
+		{name: "last", args: args{startTS: 20, num: 10}, wantRef: &type1, wantDelta: []int{}},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, got1, err := LoadByEventIdWithDelta(db, dbEvent.ID, tt.args.startTS, tt.args.num)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("LoadByEventIdWithDelta() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != nil {
+				assert.Equal(t, tt.wantRef, &got.Type)
+			} else {
+				assert.Equal(t, tt.wantRef, nil)
+			}
+			rest := make([]int, 0)
+			for _, v := range got1 {
+				rest = append(rest, v.Type)
+			}
+			assert.Equal(t, tt.wantDelta, rest)
 		})
 	}
 }

